src/typechecker.erl

```diff
@@ -1531,8 +1531,17 @@ add_type_pat_tuple(Pats, {type, _, union, Tys}, TEnv, VEnv) ->
 			    , length(TS) == length(Pats)])),
     lists:foldl(fun ({Pat, Union}, Env) ->
 			add_type_pat(Pat, Union, TEnv, Env)
-		end, VEnv, lists:zip(Pats, Unions)).
-
+                end, VEnv, lists:zip(Pats, Unions));
+add_type_pat_tuple(Pats, Ty, TEnv, VEnv)
+  when element(1, Ty) =:= user_type;
+       element(1, Ty) =:= remote_type ->
+    case normalize(Ty, TEnv) of
+        Ty ->
+            %% no change, unknown case
+            throw({unsupported_type, Ty});
```

Comment:
It would expand to itself for remote opaque types. Remote opaque types shouldn't be used in pattern matching against tuples and other decomposing patterns. Perhaps the error message can reflect that.

Reply:
good point!